# ****Data Understanding****

## Business Objective

We can say in our business we collect this dataset to see what is the progress of our business in profit and analyze different aspects of the business. So we want to see **how does revenue change over time**.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df = pd.read_csv("../data/raw/orders.csv")

df

,order_id,customer_id,order_date,ship_date,delivery_date,customer_name,customer_segment,customer_country,customer_city,customer_region,...,tax_amount,total_order_value,profit,shipping_method,delivery_status,payment_method,payment_status,sales_channel,customer_acquisition_channel,campaign
0,ORD1000000,CUST101392,2025-10-28,2025-10-28,2025-10-28,Thomas Moore,Consumer,United States,Los Angeles,North America,...,4.96,98.95,24.87,Same-Day,On Time,DEBIT CARD,Paid,Mobile App,Organic Search,NaN
1,ORD1000001,CUST117918,2025-04-22,2025-04-23,2025-04-26,Cassandra Hays,Consumer,United States,Houston,North America,...,10.96,172.91,52.96,Standard,On Time,Credit Card,Pending,Mobile App,Organic Search,NaN
2,ORD1000002,CUST111893,2025-08-01,2025-08-02,NaN,Steven Torres,Corporate,United States,Houston,North America,...,16.72,255.64,71.17,Standard,In Transit,PayPal,Paid,Marketplace,Social Media,NaN
3,ORD1000003,CUST110447,2025-11-24,2025-11-26,2025-11-28,Michelle Franco,Consumer,United States,Houston,North America,...,6.76,109.43,46.55,Standard,Delayed,Gift Card,Paid,Website,Email Marketing,BLACKFRIDAY25
4,ORD1000004,CUST102862,2025-01-15,2025-01-16,2025-01-18,Taylor Brown,Small Business,United States,New York,North America,...,3.55,60.04,14.76,Standard,On Time,Debit Card,Paid,Website,Email Marketing,LOYALTY_REWARDS
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
50120,ORD1027678,CUST132003,2024-01-15,2024-01-15,2024-01-15,Barbara Dalton,Consumer,United States,Los Angeles,North America,...,6.77,125.05,47.48,Same-Day,Delayed,Credit Card,Paid,Website,Organic Search,NaN
50121,ORD1010704,CUST101039,2024-04-02,2024-04-02,2024-04-02,Bradley Montgomery,Consumer,United States,New York,North America,...,3.62,80.55,7.13,Same-Day,On Time,Debit Card,Paid,Website,Organic Search,NaN
50122,ORD1033461,CUST128428,2026-04-25,2026-04-26,2026-04-29,Christina Swanson,Consumer,United States,Los Angeles,North America,...,6.31,103.85,40.24,Standard,On Time,PayPal,Paid,Website,Organic Search,NaN
50123,ORD1022001,CUST105020,2024-02-04,2024-02-04,2024-02-09,Tammy Booker,Corporate,Canada,Vancouver,North America,...,25.56,222.14,58.81,Standard,On Time,PayPal,Paid,Website,Organic Search,NaN


In [3]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 50125 entries, 0 to 50124
Data columns (total 27 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   order_id                      50125 non-null  str    
 1   customer_id                   50125 non-null  str    
 2   order_date                    50125 non-null  str    
 3   ship_date                     41577 non-null  str    
 4   delivery_date                 36691 non-null  str    
 5   customer_name                 50125 non-null  str    
 6   customer_segment              50125 non-null  str    
 7   customer_country              50125 non-null  str    
 8   customer_city                 49123 non-null  str    
 9   customer_region               49373 non-null  str    
 10  order_status                  50125 non-null  str    
 11  total_items                   50125 non-null  int64  
 12  unique_products               50125 non-null  int64  
 13  product_cate

In [5]:
df.shape

(50125, 27)

The primary key in this dataset could be `order_key` that we can generate for the dataset and the business key is `order_id`.

## Relationships

`customer_id` is the foreign key for the orders table and relate customers to orders.

### **Categorical fields**

- order_id
- customer_id
- customer_name
- customer_segment
- customer_country
- customer_city
- customer_region
- order_status
- product_categories
- shipping_method
- dellivery_status
- payment_method
- payment_status
- sales_channel
- customer_acquisition_channel
- campaign

### **Numerical fields**

- total_items
- unique_products
- subtotal
- dicount_amount
- shipping_cost
- tax_amount
- total_order_value
- profit

### **Date/time fields**

- order_date
- ship_date
- delivery_date

### **Measures vs. Dimensions**

> **Dimension** describes something.

> **Measure** quantifies something.
 
- order_id 	                   --> **dimension**
- customer_id                  --> **dimension**
- order_date                   --> **dimension**
- ship_date                    --> **dimension**
- delivery_date                --> **dimension**
- customer_name                --> **dimension**
- customer_segment             --> **dimension**
- customer_country             --> **dimension**
- customer_city                --> **dimension**
- customer_region              --> **dimension**
- order_status                 --> **dimension**
- total_items                  --> `measure`
- unique_products              --> `measure`
- product_categories           --> **dimension**
- subtotal                     --> `measure`
- discount_amount              --> `measure`
- shipping_cost                --> `measure`
- tax_amount                   --> `measure`
- total_order_value            --> `measure`
- profit                       --> `measure`
- shipping_method              --> **dimension**
- delivery_status              --> **dimension**
- payment_method               --> **dimension**
- payment_status               --> **dimension**
- sales_channel                --> **dimension**
- customer_acquisition_channel --> **dimension**
- campaign                     --> **dimension**

In [9]:
df.isnull().sum()

order_id                            0
customer_id                         0
order_date                          0
ship_date                        8548
delivery_date                   13434
customer_name                       0
customer_segment                    0
customer_country                    0
customer_city                    1002
customer_region                   752
order_status                        0
total_items                         0
unique_products                     0
product_categories                  0
subtotal                            0
discount_amount                  1004
shipping_cost                     602
tax_amount                          0
total_order_value                   0
profit                              0
shipping_method                     0
delivery_status                   501
payment_method                    401
payment_status                      0
sales_channel                       0
customer_acquisition_channel        0
campaign    

In [27]:
df[[
    'ship_date', 'delivery_date', 'order_status', 'delivery_status'
]]

,ship_date,delivery_date,order_status,delivery_status
0,2025-10-28,2025-10-28,Delivered,On Time
1,2025-04-23,2025-04-26,Delivered,On Time
2,2025-08-02,NaN,Shipped,In Transit
3,2025-11-26,2025-11-28,Delivered,Delayed
4,2025-01-16,2025-01-18,Delivered,On Time
...,...,...,...,...
50120,2024-01-15,2024-01-15,Delivered,Delayed
50121,2024-04-02,2024-04-02,Delivered,On Time
50122,2026-04-26,2026-04-29,Delivered,On Time
50123,2024-02-04,2024-02-09,Delivered,On Time


In [31]:
df[[
    'customer_country', 'customer_city', 'customer_region'
]][df['customer_region'].isnull()]['customer_country'].value_counts()

customer_country
United States     430
United Kingdom    114
Canada            102
Germany            76
Australia          22
USA                 8
Name: count, dtype: int64

Potential Data-quality risks
===========================
- USA & United States both are available in the dataset and needs to be fixed.
- There are 752 customer regions that are missing and we can impute them using the customer_country column.
- There are 1,002 customer_city that are missing and we can impute them using the customer_region & costomer_country columns.